controllers/volumereplicationgroup_controller.go

```diff
@@ -898,15 +898,98 @@ func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName, state volr
 			Namespace: vrNamespacedName.Namespace,
 		},
 		Spec: volrep.VolumeReplicationSpec{
-			DataSource: &corev1.TypedLocalObjectReference{
+			DataSource: corev1.TypedLocalObjectReference{
 				Kind: "PersistentVolumeClaim",
 				Name: vrNamespacedName.Name,
 			},
-			State: state,
+			ReplicationState: state,
+
+			VolumeReplicationClass: v.instance.Spec.VolumeReplicationClass,
 		},
 	}
 
-	return v.reconciler.Create(v.ctx, cr)
+	// Let VRG receive notification for any changes to VolumeReplication CR
+	// created by VRG.
+	if err := ctrl.SetControllerReference(v.instance, cr, v.reconciler.Scheme); err != nil {
+		return fmt.Errorf("failed to set owner reference to VolumeReplication resource (%s/%s)",
+			cr.Name, cr.Namespace)
+	}
+
+	v.log.Info("Creating VolumeReplication resource", "resource", cr)
+
+	if err := v.reconciler.Create(v.ctx, cr); err != nil {
+		return fmt.Errorf("failed to create VolumeReplication resource (%s)", vrNamespacedName)
+	}
+
+	return nil
+}
+
+func (v *VRGInstance) checkVRStatus(volRep *volrep.VolumeReplication) error {
+	// When the generation in the status is updated, VRG would get a reconcile
+	// as it owns VolumeReplication resource.
+	if volRep.Generation != volRep.Status.ObservedGeneration {
+		v.log.Info("Generation from the resource and status not same")
+
+		return nil
+	}
+
+	switch {
+	case v.instance.Spec.ReplicationState == volrep.Primary:
+		return v.validateVRStatusAsPrimary(volRep)
+	case v.instance.Spec.ReplicationState == volrep.Secondary:
+		return v.validateVRStatusAsSecondary(volRep)
+	case v.instance.Spec.ReplicationState == volrep.Resync:
+		return fmt.Errorf("resync is not supported for VRG")
```

Comment:
It may be safer to prevent the user from setting the state to Resync to beginwith instead of throwing an error later.  This may require us to not reuse volRep's state definition and define our own.

On the otherhand, we don't seem to be importing ReplicationClass!

Reply:
I would prefer VRG to have its own state definitions as well. Based on what @raghavendrabhat thinks, we can possibly open an issue and tackle it subsequently?

Reply:
I agree that it might be better for VRG to have its own state definitions. But, another thought I have is, since the state of VRG is very dependent on the states that VolRep has, is it better to have VolRep not export the states that it think is more specific to itself? For example, if VolRep CR does not export "resync" state and keeps that state as private to itself, then VRG can use only the exported states of VolRep.

The reason being, currently, except for "resync" VRG and VolRep agree on "primary" and "secondary" being the other states. In future, if VolRep changes itself and say uses multiple variables to decide whether primary or secondary (instead of using a single variable as of now), then VRG should have to make changes to VolRep creation code every time VolRep makes changes to its understanding of states.

Reply:
> The reason being, currently, except for "resync" VRG and VolRep agree on "primary" and "secondary" being the other states. In future, if VolRep changes itself and say uses multiple variables to decide whether primary or secondary (instead of using a single variable as of now), then VRG should have to make changes to VolRep creation code every time VolRep makes changes to its understanding of states.

The above is a good reason to keep them separate. VRG primary/secondary interface need not change/adapt to VR interface (and hence expose it to the user as well). The code can take care of the mappings and extensions as required.

Reply:
ACK. Will be doing it in a separate PR.